components/scream/src/physics/p3/p3_functions_cloud_sed_impl.hpp

```diff
@@ -0,0 +1,138 @@
+#ifndef P3_FUNCTIONS_CLOUD_SED_IMPL_HPP
+#define P3_FUNCTIONS_CLOUD_SED_IMPL_HPP
+
+#include "p3_functions.hpp"
+
+namespace scream {
+namespace p3 {
+
+/*
+ * Implementation of p3 cloud sedimentation function. Clients should NOT #include
+ * this file, #include p3_functions.hpp instead.
+ */
+
+template <typename S, typename D>
+KOKKOS_FUNCTION
+void Functions<S,D>
+::cloud_sedimentation(
+    const uview_1d<const Spack>& qc_incld,
+    const uview_1d<const Spack>& rho,
+    const uview_1d<const Spack>& inv_rho,
+    const uview_1d<const Spack>& lcldm,
+    const uview_1d<const Spack>& acn,
+    const uview_1d<const Spack>& inv_dzq,
+    const view_dnu_table& dnu,
+    const MemberType& team,
+    const Workspace& workspace,
+    const Int& nk, const Int& ktop, const Int& kbot, const Int& kdir, const Scalar& dt, const Scalar& odt, const bool& log_predictNc,
+    const uview_1d<Spack>& qc,
+    const uview_1d<Spack>& nc,
+    const uview_1d<Spack>& nc_incld,
+    const uview_1d<Spack>& mu_c,
+    const uview_1d<Spack>& lamc,
+    const uview_1d<Spack>& qc_tend,
+    const uview_1d<Spack>& nc_tend,
+    Scalar& prt_liq)
+{
+  // Get temporary workspaces needed for the cloud-sed calculation
+  uview_1d<Spack> V_qc, V_nc, flux_qx, flux_nx;
+  workspace.template take_many<4>(
+    {"V_qc", "V_nc", "flux_qx", "flux_nx"},
+    {&V_qc, &V_nc, &flux_qx, &flux_nx});
+
+  view_1d_ptr_array<Spack, 2>
+    fluxes_ptr = {&flux_qx, &flux_nx},
+    vs_ptr     = {&V_qc, &V_nc},
+    qnr_ptr    = {&qc, &nc};
+
+  view_1d_ptr_array<Spack, 1>
+    flux_ptr  = {&flux_qx},
+    v_ptr     = {&V_qc},
+    qr_ptr    = {&qc};
+
+  // find top, determine qxpresent
+  const auto sqc = scalarize(qc);
+  constexpr Scalar qsmall = C::QSMALL;
+  bool log_qxpresent;
+  const Int k_qxtop = find_top(team, sqc, qsmall, kbot, ktop, kdir, log_qxpresent);
+
+  if (log_qxpresent) {
+    Scalar dt_left = dt;    // time remaining for sedi over full model (mp) time step
+    Scalar prt_accum = 0.0; // precip rate for individual category
+
+    // find bottom
+    Int k_qxbot = find_bottom(team, sqc, qsmall, kbot, k_qxtop, kdir, log_qxpresent);
+
+    while (dt_left > C::dt_left_tol) {
+      Scalar Co_max = 0.0;
+      Int kmin, kmax;
+      const Int kmin_scalar = ( kdir == 1 ? k_qxbot : k_qxtop);
+      const Int kmax_scalar = ( kdir == 1 ? k_qxtop : k_qxbot);
+
+      Kokkos::parallel_for(
+        Kokkos::TeamThreadRange(team, V_qc.extent(0)), [&] (Int k) {
+          V_qc(k) = 0;
+          if (log_predictNc) {
+            V_nc(k) = 0;
+          }
+      });
+      team.team_barrier();
+
+      // Convert top/bot to pack indices
+      util::set_min_max(k_qxbot, k_qxtop, kmin, kmax, Spack::n);
+
+      Kokkos::parallel_reduce(
+        Kokkos::TeamThreadRange(team, kmax-kmin+1), [&] (int pk_, Scalar& lmax) {
+          const int pk = kmin + pk_;
+          const auto range_pack = scream::pack::range<IntSmallPack>(pk*Spack::n);
```

Comment:
One tricky thing was that, in single precision, some of values outside of the active k-range (qxbot-qxtop) had qc_incld(k)>qsmall. I had to add this range mask logic to make sure these levels were not processed by get_cloud_dsd2.